### PR TITLE
USB: Add MS OS Descriptors version 1 support

### PR DESCRIPTION
**Pull request by finikorg**

### CODE REVIEW

**Comment by finikorg** [comment]
@IvanSanchez

**Comment by carlescufi** [comment]
@jfischer-phytec-iot @pawelzadrozniak any feedback on this one?

**Comment by jfischer-phytec-iot** [comment]
will do, just have to find time

**Comment by jfischer-phytec-iot** [comment]
@finikorg please rebase

**Comment by finikorg** [comment]
@jfischer-phytec-iot rebased

File: subsys/usb/usb_device.c [hunk 1]
```diff
@@ -898,12 +922,19 @@ int usb_set_config(struct usb_cfg_data *config)
 					     config->interface.class_handler,
 					     config->interface.payload_data);
 	}
-	/* register vendor request handlers */
-	if (config->interface.vendor_handler) {
+
+	/* register vendor request handler */
+	if (config->interface.vendor_handler || usb_os_desc_enabled()) {
```
Comment (jfischer-phytec-iot) [inline subsys/usb/usb_device.c]: There is already a wrapper for composite configuration:
https://github.com/zephyrproject-rtos/zephyr/pull/7916/commits/b09e5dcb0e56f5ca6800938fdb27a91dbc194e8b#diff-609952c961f3c77cb7811bedee4bfd2cL1388

That also needs to be changed:
https://github.com/zephyrproject-rtos/zephyr/pull/7916/commits/b09e5dcb0e56f5ca6800938fdb27a91dbc194e8b#diff-609952c961f3c77cb7811bedee4bfd2cR1494

We can simplify it all and only use usb_composite_init for initialization without explicitly turning on CONFIG_USB_COMPOSITE_DEVICE, but first interface descriptors in samples webusb and wpanusb have to be reworked.

Comment (carlescufi) [reply subsys/usb/usb_device.c]: @finikorg any comments on this? I think this makes sense

Comment (finikorg) [reply subsys/usb/usb_device.c]: @jfischer-phytec-iot @carlescufi vendor_handler() in usb_devices.c does not handle request correctly. It expects that setup->wIndex indicates interface number, which is not the case for some (maybe all) of requests. 
"Microsoft OS 2.0 Descriptors Specification":
```
wIndex 0x7 for MS_OS_20_DESCRIPTOR_INDEX.
wIndex 0x8 for MS_OS_20_SET_ALT_ENUMERATION
...
etc
```

Comment (finikorg) [reply subsys/usb/usb_device.c]: @jfischer-phytec-iot Added patch #afe05be

Comment (finikorg) [reply subsys/usb/usb_device.c]: @jfischer-phytec-iot Please have a look at 96a9583

File: subsys/usb/os_desc.c [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		       s32_t *len, u8_t **data)
+{
+	SYS_LOG_DBG("wValue 0x%x", setup->wValue);
+
+	if (!os_desc) {
+		return -ENOTSUP;
+	}
+
+	if (GET_DESC_TYPE(setup->wValue) == DESC_STRING &&
+	    GET_DESC_INDEX(setup->wValue) == 0xEE) {
```
Comment (jfischer-phytec-iot) [inline subsys/usb/os_desc.c]: Can 0xEE be replaced by a macro or otherwise documented?

Comment (finikorg) [reply subsys/usb/os_desc.c]: This is possible, I will add documentation block.

**Review by jfischer-no — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**